### PR TITLE
ITKDev: Entity audit module

### PR DESCRIPTION
**Pull request by cableman**
#### Link to ticket

https://leantime.itkdev.dk/#/tickets/showTicket/3206

#### Description

* Add audit check of entities, mostly to check web-form submissions.

#### Screenshot of the result

N/A

#### Checklist

- [x] My code passes our static analysis suite.
- [x] My code passes our continuous integration process.

If your code does not pass all the requirements on the checklist you have to add a comment explaining why this change
should be exempt from the list.

#### Additional comments or questions

N/A

### CODE REVIEW

File: modules/os2web_audit_entity/os2web_audit_entity.module [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+    $personal = '';
+    $submissionData = $submission->getData();
+    $filterFields = preg_grep('/^(.*)cpr(.*)$/', array_keys($submissionData));
+    if (!empty($filterFields)) {
+      foreach ($filterFields as $field) {
+        $cpr = $submissionData[$field];
+        $personal .= sprintf('CPR "%s".', $cpr);
+      }
+    }
+
```
Comment (jekuaitk) [inline modules/os2web_audit_entity/os2web_audit_entity.module:68]: Are we interested in logging submission access if `$personal` is empty?

File: modules/os2web_audit_entity/os2web_audit_entity.module [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+  // Detect user type.
+  $account = \Drupal::currentUser();
+  $apiUser = FALSE;
+  if (os2web_audit_entity_is_api_user($account)) {
+    $apiUser = TRUE;
+  }
+
+  $metadata['userId'] = $account->getEmail();
+  $metadata['userType'] = $apiUser ? 'api' : 'web';
+  $logger->info('Entity', $message, FALSE, $metadata);
```
Comment (jekuaitk) [inline modules/os2web_audit_entity/os2web_audit_entity.module:124]:
```suggestion
  $metadata['userId'] = $account->getEmail();
  $metadata['userType'] = os2web_audit_entity_is_api_user($account) ? 'api' : 'web';
  $logger->info('Entity', $message, FALSE, $metadata);
```

File: modules/os2web_audit_entity/src/Form/SettingsForm.php [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+    $config = $this->config(self::$configName);
+
+    $form['roles'] = [
+      '#type' => 'checkboxes',
+      '#title' => $this->t('Choose an Option'),
+      '#description' => $this->t('Please select an option from the dropdown menu.'),
+      '#options' => $items,
+      '#default_value' => $config->get('roles') ?? [],
+      '#required' => TRUE,
+    ];
```
Comment (jekuaitk) [inline modules/os2web_audit_entity/src/Form/SettingsForm.php:78]: We should probably say that selected roles are considered API-roles and what that means with respect to logging.

**Review by jekuaitk — CHANGES_REQUESTED** [review]

File: modules/os2web_audit_entity/os2web_audit_entity.info.yml [hunk 4]
```diff
@@ -0,0 +1,6 @@
+name: "OS2web Audit logging entity access"
+description: "Logs CUD events for entities"
```
[the remark this answers is not in the document]
Comment (cableman) [reply modules/os2web_audit_entity/os2web_audit_entity.info.yml:2]: Fixed

File: modules/os2web_audit_entity/os2web_audit_entity.module (hunk 1 above)
Comment (cableman) [reply modules/os2web_audit_entity/os2web_audit_entity.module:80]: yes, "personal" is just CPR's.... but data-lak of submission with mails, address etc. still nice to known when it happened.

File: modules/os2web_audit_entity/src/Form/SettingsForm.php (hunk 3 above)
Comment (cableman) [reply modules/os2web_audit_entity/src/Form/SettingsForm.php:78]: Fixed

**Review by jekuaitk — APPROVED** [review]

**Pull request merged**